pkg/controller/replicaset/replica_set.go

```diff
@@ -650,3 +641,76 @@ func (rsc *ReplicaSetController) syncReplicaSet(key string) error {
 	}
 	return manageReplicasErr
 }
+
+// slowStartBatch obtains total batch size to be processed and a function to execute during each batch.
+// It returns successfully processed batch size and an error if found.
+func slowStartBatch(totalBatchSize int, fn func() error) (int, error) {
+	var wg sync.WaitGroup
+	remainedBatchSize := totalBatchSize
+	totalProcessedBatchSize := 0
+	errCh := make(chan error, totalBatchSize)
+	var lock sync.Mutex
+	for batchSize := integer.IntMin(remainedBatchSize, controller.SlowStartInitialBatchSize); remainedBatchSize > 0; batchSize = integer.IntMin(2*batchSize, remainedBatchSize) {
+		wg.Add(batchSize)
+		for i := 0; i < batchSize; i++ {
+			go func() {
+				defer wg.Done()
+				err := fn()
+				if err != nil && !errors.IsTimeout(err) {
+					errCh <- err
+				} else if err == nil {
+					lock.Lock()
+					totalProcessedBatchSize++
+					lock.Unlock()
+				}
+
+			}()
+		}
+		wg.Wait()
+		remainedBatchSize -= batchSize
+	}
+
+	select {
+	case err := <-errCh:
+		// all errors have been reported before and they're likely to be the same, so we'll only return the first one we hit.
+		if err != nil {
+			return totalProcessedBatchSize, err
+		}
+	default:
+	}
+	return totalProcessedBatchSize, nil
+}
+
+func getCanAdoptFunc(rsc *ReplicaSetController, rs *extensions.ReplicaSet) func() error {
+	// If any adoptions are attempted, we should first recheck for deletion with
+	// an uncached quorum read sometime after listing Pods (see #42639).
+	return controller.RecheckDeletionTimestamp(func() (metav1.Object, error) {
+		fresh, err := rsc.kubeClient.ExtensionsV1beta1().ReplicaSets(rs.Namespace).Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		if fresh.UID != rs.UID {
+			return nil, fmt.Errorf("original ReplicaSet %v/%v is gone: got uid %v, wanted %v", rs.Namespace, rs.Name, fresh.UID, rs.UID)
+		}
+		return fresh, nil
+	})
+}
+
+func getPodsToDeleteDetails(filteredPods []*v1.Pod, diff int) ([]string, []*v1.Pod) {
+	// Snapshot the UIDs (ns/name) of the pods we're expecting to see
+	// deleted, so we know to record their expectations exactly once either
+	// when we see it as an update of the deletion timestamp, or as a delete.
+	// Note that if the labels on a pod/rs change in a way that the pod gets
+	// orphaned, the rs will only wake up after the expectations have
+	// expired even if other pods are deleted.
+	podsToDeleteKeys := []string{}
```

Comment:
Apart from pods to delete, we can also refactor out the codes for getting pod keys. A loop is sufficient to get both the pod keys and pods.